src/main/java/com/google/devtools/build/lib/analysis/constraints/RuleContextConstraintSemantics.java

```diff
@@ -808,4 +825,147 @@ private static void addSelectValuesToSet(BuildType.Selector<?> select, final Set
       type.visitLabels(visitor, value, /*context=*/ null);
     }
   }
+
+  /**
+   * Creates an incompatible {@link ConfiguredTarget} if the corresponding rule is incompatible.
+   *
+   * <p>Returns null if the target is not incompatible.
+   *
+   * <p>"Incompatible" here means either:
+   *
+   * <ol>
+   *   <li>the corresponding <code>target_compatible_with</code> list contains a constraint that the
+   *       current platform doesn't satisfy, or
+   *   <li>a transitive dependency is incompatible.
+   * </ol>
+   *
+   * @param ruleContext analysis context for the rule
+   * @param prerequisiteMap the dependencies of the rule
+   * @throws ActionConflictException if the underlying {@link RuleConfiguredTargetBuilder}
+   *     encounters a problem when assembling a dummy action for the incompatible {@link
+   *     ConfiguredTarget}.
+   */
+  @Nullable
+  public static ConfiguredTarget incompatibleConfiguredTarget(
+      RuleContext ruleContext,
+      OrderedSetMultimap<DependencyKind, ConfiguredTargetAndData> prerequisiteMap)
+      throws ActionConflictException {
+    if ("toolchain".equals(ruleContext.getRule().getRuleClass())) {
+      return null;
+    }
+
+    // This is incompatible if explicitly specified to be.
+    if (ruleContext.attributes().has("target_compatible_with")) {
+      Iterable<ConstraintValueInfo> constraintValues =
+          PlatformProviderUtils.constraintValues(
+              ruleContext.getPrerequisites("target_compatible_with"));
+      for (ConstraintValueInfo constraintValue : constraintValues) {
+        if (!ruleContext.targetPlatformHasConstraint(constraintValue)) {
+          return createIncompatibleConfiguredTarget(ruleContext, null, constraintValue);
```

Comment:
Related to my earlier comment, we should report all constraint values that are in violation, not just the first. Same thing for the next check, below.

If we are collecting all of these, a stream might be easier to express:
ImmutableList<ConstraintValue> incompatibleConstraintValues =
    PlatformProviderUtils.constraintValues(
              ruleContext.getPrerequisites("target_compatible_with")).stream()
    .filter(cv -> !ruleContext.targetPlatformHasConstraint(cv)
    .collect(ImmutableList.toImmutableList());

Similarly for the dependencies below.

Reply:
Nice, done.  `PlatformProviderUtils.constraintValues(
ruleContext.getPrerequisites("target_compatible_with"))` returns an Iterable which doesn't have a stream method.  I poked around a bit and couldn't find a reasonable adapter to do that from other examples.  Did I miss something?

Reply:
Oh, use conm.google.common.collect.Streams.stream. At some point I should rework PlatformProviderUtils to return actual Collections, they are underneath.

Reply:
Thanks!  Done.